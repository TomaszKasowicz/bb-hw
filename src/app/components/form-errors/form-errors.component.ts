import { Component, Input, OnInit } from '@angular/core';
import { AbstractControl } from '@angular/forms';

export interface ErrorMessages {
  [key: string]: string;
}

@Component({
  selector: 'app-form-errors',
  template: `
    <ng-container *ngIf="ctrl?.dirty && ctrl?.errors">
      <div class="error" *ngFor="let error of ctrl?.errors | keyvalue">
        {{ errorMessages[error.key] || error.key }}
      </div>
    </ng-container>
  `,
  styles: [
    `
      :host {
        width: 100%;
      }

      .error {
        font-size: 0.75em;
        width: 100%;
        color: var(--color-peachtree-orange);
      }
    `
  ]
})
export class FormErrorsComponent {
  @Input()
  public ctrl: AbstractControl | null;

  @Input()
  public errorMessages: ErrorMessages = {};
}
