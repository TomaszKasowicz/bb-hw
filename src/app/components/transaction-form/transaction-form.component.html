<app-window-header 
  [icon]="'/assets/icons/arrows.png'">
  Make a Transfer
</app-window-header>
<div class="flex flex_column transfer-form" [formGroup]="transferForm">
  <label class="font__bold" for="from">FROM ACCOUNT</label>
  <input type="text" formControlName="from">
  <label class="font__bold"for="to">TO ACCOUNT</label>
  <input type="text" formControlName="to">
  <label class="font__bold" for="amount">AMOUNT</label>
  <div class="amount">
    <span class="prefix">$</span>    
    <input type="number" formControlName="amount">
  </div>
  <div class="flex buttons">
    <ng-container *ngIf="isReview === false; else review">
      <button class="peachtree"
      [disabled]="disabled$ | async" 
      (click)="toggleState()">SUBMIT</button>  
    </ng-container>
  </div>
  <ng-template #review>
      <button class="peachtree" (click)="toggleState()">CANCEL</button>
      <button class="peachtree" (click)="transfer()">TRANSFER</button>  
  </ng-template>
  
</div>