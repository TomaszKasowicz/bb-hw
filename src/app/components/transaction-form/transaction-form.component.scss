.transfer-form {
  padding: 1em;

  input,
  label {
    width: 100%;
  }

  input {
    margin-bottom: 1em;
    padding: 0.5em 0em;
  }

  .amount {
    width: 100%;

    input {
      padding-left: 15px;
    }
  }

  .buttons {
    width: 100%;
    justify-content: flex-end;

    button {
      flex-basis: 10em;
    }
  }
}

.prefix {
  color: var(--color-text-disabled);
  position: absolute;
}
