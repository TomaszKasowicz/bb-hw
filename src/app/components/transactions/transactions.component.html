<app-transaction-form class="window flex-grow-1"></app-transaction-form>
<app-recent-transactions class="window flex-grow-2"></app-recent-transactions>
