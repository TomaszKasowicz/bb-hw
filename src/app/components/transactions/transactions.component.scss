:host {
  display: flex;
  justify-content: space-between;
  padding: 5em;
  gap: 4em;
}
