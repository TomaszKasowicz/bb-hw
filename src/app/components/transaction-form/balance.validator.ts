import { AbstractControl, ValidationErrors } from '@angular/forms';
import { TransactionsService } from '../../services/transactions.service';

const MIN_BALANCE = -500.0;

export function balanceValidator(transactionsService: TransactionsService) {
  return (control: AbstractControl): ValidationErrors | null => {
    const amount = +control.value;
    const balance = transactionsService.getCurrentBalance();
    const newBalance = balance - amount;

    if (newBalance < MIN_BALANCE) {
      return { 'balance-overdraft': true };
    }
    return null;
  };
}
