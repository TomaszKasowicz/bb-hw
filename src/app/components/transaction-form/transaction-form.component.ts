import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { defer, Observable } from 'rxjs';
import { map, startWith, tap } from 'rxjs/operators';
import { Transaction } from '../../model/transaction';
import { TransactionsService } from '../../services/transactions.service';
import { balanceValidator } from './balance.validator';

@Component({
  selector: 'app-transaction-form',
  templateUrl: './transaction-form.component.html',
  styleUrls: ['./transaction-form.component.scss']
})
export class TransactionFormComponent implements OnInit {
  public transferForm: FormGroup;
  public isReview: boolean;

  public disabled$: Observable<boolean>;

  constructor(
    private fb: FormBuilder,
    private transactionService: TransactionsService
  ) {
    this.transferForm = this.fb.group({
      from: this.fb.control({ value: null, disabled: true }),
      to: this.fb.control(null, [Validators.required]),
      amount: this.fb.control(null, [
        Validators.required,
        Validators.min(0),
        balanceValidator(this.transactionService)
      ])
    });
  }

  private resetForm() {
    this.isReview = false;
    this.transferForm.reset({
      from: this.getFromValue(),
      to: null,
      amount: null
    });
    this.transferForm.enable();
    this.transferForm.get('from')?.disable();
    // this.transferForm.updateValueAndValidity();
  }

  ngOnInit(): void {
    this.resetForm();
    this.disabled$ = defer(() => this.transferForm.statusChanges).pipe(
      map(status => status === 'INVALID'),
      startWith(true)
    );
  }

  private getFromValue() {
    return `Free Checking(4962) - $${Number(
      this.transactionService.getCurrentBalance()
    ).toFixed(2)}`;
  }

  toggleState() {
    this.isReview = !this.isReview;
    if (!this.isReview) {
      this.resetForm();
    } else {
      this.transferForm.disable();
    }
  }

  transfer() {
    const value = this.transferForm.getRawValue();
    const transaction: Transaction = {
      amount: value.amount,
      categoryCode: '#1180aa',
      merchant: value.to,
      transactionType: 'Transaction',
      transactionDate: new Date().getTime()
    };
    this.transactionService.submitTransaction(transaction);
    this.resetForm();
  }
}
