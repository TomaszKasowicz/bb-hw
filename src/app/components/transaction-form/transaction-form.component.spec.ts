import { ReactiveFormsModule } from '@angular/forms';
import {
  createComponentFactory,
  mockProvider,
  Spectator,
  byText
} from '@ngneat/spectator/jest';
import { of } from 'rxjs';
import { TransactionsService } from '../../services/transactions.service';
import { FormErrorsComponent } from '../form-errors/form-errors.component';
import { TransactionFormComponent } from './transaction-form.component';

describe('TransactionFormComponent', () => {
  let spectator: Spectator<TransactionFormComponent>;
  const createComponent = createComponentFactory({
    component: TransactionFormComponent,
    declarations: [TransactionFormComponent, FormErrorsComponent],
    imports: [ReactiveFormsModule],
    providers: [mockProvider(TransactionsService)],
    shallow: true,
    detectChanges: false
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('Should render clean form without any options', () => {
    const transactionsService = spectator.inject<TransactionsService>(
      TransactionsService
    );
    transactionsService.getRecentTransactions.mockReturnValue(of([]));
    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();

    const submitButton = spectator.query(byText('SUBMIT'));
    expect(submitButton).toBeTruthy();
    expect(submitButton?.getAttribute('disabled')).toEqual('');
  });

  it('Should Render Merchant options retrieved from recent transactions', () => {
    const transactionsService = spectator.inject<TransactionsService>(
      TransactionsService
    );
    const transactions: any[] = [
      { merchant: 'a' },
      { merchant: 'b' },
      { merchant: 'b' }
    ];
    transactionsService.getRecentTransactions.mockReturnValue(of(transactions));
    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();

    const submitButton = spectator.query(byText('SUBMIT'));
    expect(submitButton).toBeTruthy();
    expect(submitButton?.getAttribute('disabled')).toEqual('');

    const options = spectator.queryAll('option');
    expect(options.length).toEqual(transactions.length - 1);
  });

  describe('Form Submission', () => {
    const merchant = 'merchant';
    const amount = 123.45;

    beforeEach(() => {
      const transactionsService = spectator.inject<TransactionsService>(
        TransactionsService
      );
      transactionsService.getRecentTransactions.mockReturnValue(of([]));
      spectator.detectChanges();

      spectator.typeInElement(merchant, '#to');
      spectator.typeInElement(`${amount}`, '#amount');
      spectator.detectChanges();
    });

    it('Should Enable Submit Button when beneficiary is filled and ammount is filled', () => {
      expect(spectator.fixture).toMatchSnapshot();

      const submitButton = spectator.query(byText('SUBMIT'));
      expect(submitButton).toBeTruthy();
      expect(submitButton?.getAttribute('disabled')).toBeNull();
    });

    it('Should disable form controls and render cancel/transfer buttons when submit clicked', () => {
      spectator.click(byText('SUBMIT'));
      spectator.detectChanges();

      expect(spectator.fixture).toMatchSnapshot();

      const buttons = spectator.queryAll('button');
      expect(buttons.length).toEqual(2);

      const inputToElement = spectator.query('#to');
      const inputAmountElement = spectator.query('#amount');
      expect(inputToElement?.getAttribute('disabled')).toEqual('');
      expect(inputAmountElement?.getAttribute('disabled')).toEqual('');
    });

    it('Should Send Transfer when transfer button is clicked', () => {
      spectator.click(byText('SUBMIT'));
      spectator.detectChanges();
      spectator.click(byText('TRANSFER'));
      spectator.detectChanges();

      expect(spectator.fixture).toMatchSnapshot();

      expectSubmitButtonIsDisabled();
      expectFormIsCleared();

      const transactionService = spectator.inject(TransactionsService);
      expect(transactionService.submitTransaction).toHaveBeenCalledTimes(1);
      expect(transactionService.submitTransaction).toHaveBeenCalledWith(
        expect.objectContaining({
          amount: amount,
          categoryCode: '#1180aa',
          merchant: merchant,
          transactionType: 'Transaction'
        })
      );
    });

    it('Should Reset Form when cancel is clicked', () => {
      spectator.click(byText('SUBMIT'));
      spectator.detectChanges();
      spectator.click(byText('CANCEL'));
      spectator.detectChanges();

      expect(spectator.fixture).toMatchSnapshot();
      expectSubmitButtonIsDisabled();
      expectFormIsCleared();

      const transactionService = spectator.inject(TransactionsService);
      expect(transactionService.submitTransaction).toHaveBeenCalledTimes(0);
    });
  });

  describe('Form Errors', () => {
    beforeEach(() => {
      const transactionsService = spectator.inject<TransactionsService>(
        TransactionsService
      );
      transactionsService.getRecentTransactions.mockReturnValue(of([]));
      spectator.detectChanges();
    });

    it('Should render that Beneficiary is required', () => {
      spectator.typeInElement('Beneficiary', '#to');
      spectator.detectChanges();
      spectator.typeInElement('', '#to');

      expect(spectator.fixture).toMatchSnapshot();
      expect(spectator.query(byText('Beneficiary is required'))).toBeTruthy();
    });

    it('Should render that Amount is required', () => {
      spectator.typeInElement('123', '#amount');
      spectator.detectChanges();
      spectator.typeInElement('', '#amount');

      expect(spectator.fixture).toMatchSnapshot();
      expect(spectator.query(byText('Amount is required'))).toBeTruthy();
    });

    it('Should render that Amount should be a fixed number', () => {
      spectator.typeInElement('abcdefg', '#amount');
      spectator.detectChanges();

      expect(spectator.fixture).toMatchSnapshot();
      expect(
        spectator.query(byText('Fixed number (i.e. 123 or 123.45)'))
      ).toBeTruthy();
    });

    it('Should render that Amount will overdraft account balance', () => {
      const service = spectator.inject(TransactionsService);
      service.getCurrentBalance.mockReturnValue(400);

      spectator.typeInElement('1000', '#amount');
      spectator.detectChanges();

      expect(spectator.fixture).toMatchSnapshot();
      expect(spectator.query(byText('Will overdraft'))).toBeTruthy();
    });
  });

  function expectSubmitButtonIsDisabled() {
    const submitButton = spectator.query(byText('SUBMIT'));
    expect(submitButton).toBeTruthy();
    expect(submitButton?.getAttribute('disabled')).toEqual('');
  }

  function expectFormIsCleared() {
    const inputFromElement = spectator.query('#from');
    const inputToElement = spectator.query('#to');
    const inputAmountElement = spectator.query('#amount');
    expect(inputFromElement?.getAttribute('disabled')).toEqual('');
    expect(inputToElement?.getAttribute('disabled')).toEqual(null);
    expect(inputAmountElement?.getAttribute('disabled')).toEqual(null);
  }
});
