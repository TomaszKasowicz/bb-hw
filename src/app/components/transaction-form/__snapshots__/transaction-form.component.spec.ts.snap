// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionFormComponent Form Errors Should render that Amount is required 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-invalid ng-dirty"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      <div
        class="error"
      >
         Amount is required 
      </div>
      
      
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-dirty ng-invalid"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Form Errors Should render that Amount should be a fixed number 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-invalid ng-dirty"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      <div
        class="error"
      >
         Fixed number (i.e. 123 or 123.45) 
      </div>
      
      
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-invalid ng-dirty"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Form Errors Should render that Amount will overdraft account balance 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-invalid ng-dirty"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      <div
        class="error"
      >
         Will overdraft 
      </div>
      
      
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-invalid ng-dirty"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Form Errors Should render that Beneficiary is required 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-invalid ng-dirty"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      <div
        class="error"
      >
         Beneficiary is required 
      </div>
      
      
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-dirty ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-pristine ng-invalid"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Form Submission Should Enable Submit Button when beneficiary is filled and ammount is filled 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-dirty ng-touched ng-valid"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-dirty ng-valid ng-touched"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-dirty ng-valid"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Form Submission Should Reset Form when cancel is clicked 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-pristine ng-invalid"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Form Submission Should Send Transfer when transfer button is clicked 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-pristine ng-invalid"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Form Submission Should disable form controls and render cancel/transfer buttons when submit clicked 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview={[Function Boolean]}
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-dirty ng-touched"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-dirty ng-touched"
      disabled=""
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-dirty"
        disabled=""
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
      >
        CANCEL
      </button>
      <button
        class="peachtree"
      >
        TRANSFER
      </button>
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Should Render Merchant options retrieved from recent transactions 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      <option>
        a
      </option>
      <option>
        b
      </option>
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-pristine ng-invalid"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;

exports[`TransactionFormComponent Should render clean form without any options 1`] = `
<app-transaction-form
  __ngContext__={[Function LRootView]}
  amountErrorMessages={[Function Object]}
  beneficiaryErrroMessages={[Function Object]}
  ctrlAmount={[Function FormControl]}
  ctrlTo={[Function FormControl]}
  disabled$={[Function Observable]}
  fb={[Function FormBuilder]}
  isReview="false"
  merchants$={[Function Observable]}
  transactionService={[Function Object]}
  transferForm={[Function FormGroup]}
>
  <app-window-header>
     Make a Transfer
  
  </app-window-header><form
    class="flex flex_column transfer-form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <label
      class="bold"
      for="from"
    >
      FROM ACCOUNT
    </label>
    <input
      class="ng-untouched ng-pristine"
      disabled=""
      formcontrolname="from"
      id="from"
      type="text"
    />
    <label
      class="bold"
      for="to"
    >
      TO ACCOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <input
      autocomplete="off"
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="to"
      id="to"
      list="merchants"
      type="text"
    />
    <datalist
      id="merchants"
    >
      
    </datalist>
    <label
      class="bold"
      for="amount"
    >
      AMOUNT
    </label>
    <app-form-errors>
      
    </app-form-errors>
    <div
      class="amount"
    >
      <span
        class="prefix"
      >
        $
      </span>
      <input
        autocomplete="off"
        class="ng-untouched ng-pristine ng-invalid"
        formcontrolname="amount"
        id="amount"
        placeholder="0.00"
        type="text"
      />
    </div>
    <div
      class="flex buttons"
    >
      <button
        class="peachtree"
        disabled=""
      >
        SUBMIT
      </button>
      
      
    </div>
    
  </form>
</app-transaction-form>
`;
